Samples/CayenneDecoder/Classes/CayenneDecoder.cs

```diff
@@ -0,0 +1,227 @@
+            }
+            barometer = null;
+            return false;
+        }
+
+        public bool TryGetGyrometer(out Gyrometer gyrometer)
+        {
+            var ret = GetOutValues(CayenneTypes.Gyrometer);
+            if (ret != null)
+            {
+                gyrometer = new Gyrometer()
+                {
+                    Channel = ret[0],
+                    X = ((Int16)(ret[1] << 8) + ret[2]) / 100.0,
+                    Y = ((Int16)(ret[3] << 8) + ret[4]) / 100.0,
+                    Z = ((Int16)(ret[5] << 8) + ret[6]) / 100.0
+                };
+                return true;
+            }
+            gyrometer = null;
+            return false;
+        }
+
+        public bool TryGetGPSLocation(out GPSLocation gPSLocation)
+        {
+            var ret = GetOutValues(CayenneTypes.Gps);
+            if (ret != null)
+            {
+                gPSLocation = new GPSLocation();
+
+                gPSLocation.Channel = ret[0];
+                double sign = 1.0;
+                if ((ret[1] & 0x80) == 0x80)
+                {
+                    ret[1] = (byte)(ret[1] & 0x7F);
+                    sign = -1.0;
+                }
+                gPSLocation.Latitude = sign * ((ret[1] << 16) + (ret[2] << 8) + ret[3]) / 10000.0;
+                sign = 1.0;
+                if ((ret[4] & 0x80) == 0x80)
+                {
+                    ret[4] = (byte)(ret[4] & 0x7F);
+                    sign = -1.0;
+                }
+                gPSLocation.Longitude = sign * ((ret[4] << 16) + (ret[5] << 8) + ret[6]) / 10000.0;
+                sign = 1.0;
+                if ((ret[7] & 0x80) == 0x80)
+                {
+                    ret[7] = (byte)(ret[7] & 0x7F);
+                    sign = -1.0;
+                }
+                gPSLocation.Altitude = sign * ((ret[7] << 16) + (ret[8] << 8) + ret[9]) / 100.0;
+                return true;
+            }
+            gPSLocation = null;
+            return false;
+        }
+
+        private byte[] GetOutValues(CayenneTypes cayenneTypes)
+        {
+            byte[] outArray = null;
+            byte channel = 0;
+            int cursor = 0;
+            try
+            {
+                while (cursor != buffer.Length)
+                {
+                    channel = buffer[cursor++];
+                    var type = buffer[cursor++];
+                    var size = (byte)Enum.Parse<CayenneTypeSize>(Enum.GetName(typeof(CayenneTypes), type));
+                    if (type == (byte)cayenneTypes)
+                    {
+                        // found what we want to decode
+                        outArray = new byte[size - 1];
+                        outArray[0] = channel;
+                        Array.Copy(buffer, cursor, outArray, 1, outArray.Length - 1);
+                        return outArray;
+                    }
+                    cursor += size - 2;
+                }
+
+            }
+            catch (Exception)
```

Comment:
I am not sure this is a good way of tackling exception

Reply:
it's just to make sure I catch Exceptions. In this case, I jsut return a null byte array (after the loop). It's an efficient way to trap them. The function is then call by higher level functions which are TryGet pattent